authorization/msi-auth-token-provider-jar/src/main/java/com/microsoft/azure/msiAuthTokenProvider/MSICredentials.java

```diff
@@ -0,0 +1,356 @@
+        } else {
+            //We are in a vm/container
+            MSIConfigurationForVirtualMachine config = new MSIConfigurationForVirtualMachine(managementEndpoint);
+            return forVirtualMachine(config);
+        }
+    }
+
+    /**
+     * Creates MSICredentials for application running on MSI enabled virtual machine.
+     *
+     * @return MSICredentials
+     */
+    public static MSICredentials forVirtualMachine() {
+        return new MSICredentials(new MSIConfigurationForVirtualMachine());
+    }
+
+    /**
+     * Creates MSICredentials for application running on MSI enabled virtual machine.
+     *
+     * @param config the configuration to be used for token request.
+     * @return MSICredentials
+     */
+    public static MSICredentials forVirtualMachine(MSIConfigurationForVirtualMachine config) {
+        return new MSICredentials(config.clone());
+    }
+
+    /**
+     * Creates MSICredentials for application running on MSI enabled app service.
+     *
+     * @return MSICredentials
+     */
+    public static MSICredentials forAppService() {
+        return new MSICredentials(new MSIConfigurationForAppService());
+    }
+
+    /**
+     * Creates MSICredentials for application running on MSI enabled app service.
+     *
+     * @param config the configuration to be used for token request.
+     * @return MSICredentials
+     */
+    public static MSICredentials forAppService(MSIConfigurationForAppService config) {
+        return new MSICredentials(config.clone());
+    }
+
+    private MSICredentials(MSIConfigurationForVirtualMachine config) {
+        this.configForVM = config;
+        this.configForAppService = null;
+        this.hostType = HostType.VIRTUAL_MACHINE;
+        this.maxRetry = config.maxRetry() < 0 ? MAX_RETRY_DEFAULT_LIMIT : config.maxRetry();
+        // Simplified variant of https://en.wikipedia.org/wiki/Exponential_backoff
+        for (int x = 0; x < this.maxRetry; x++) {
+            this.retrySlots.add(500 * ((2 << 1) - 1) / 1000);
+        }
+    }
+
+    private MSICredentials(MSIConfigurationForAppService config) {
+        this.configForAppService = config;
+        this.configForVM = null;
+        this.hostType = HostType.APP_SERVICE;
+        this.maxRetry = -1;
+    }
+
+    /**
+     * Updates the client Id for the associated config (vm or app)
+     * Specifying a null value will clear out the old value
+     * @param clientId
+     */
+    public void updateClientId(String clientId) {
+        if (configForVM != null) {
+            configForVM.withClientId(clientId);
+        } else {
+            configForAppService.withClientId(clientId);
+        }
+    }
+
+    public Single<MSIToken> getToken(String tokenAudience) {
+        switch (hostType) {
+            case VIRTUAL_MACHINE:
+                    return Single.fromCallable(() -> this.getTokenForVirtualMachineFromIMDSEndpoint(tokenAudience == null ? this.configForVM.resource() : tokenAudience));
```

Comment:
Just a thought: We could make it async all the way down, i.e. by making `getTokenForVirtualMachineFromIMDSEndpoint` to use rx constructs that way retry-sleep is also async. May be we could keep this in mind and implement if there is a need.

Reply:
I am thinking of totally doing away with the rx dependency of this library... its meant to be a really light weight tar with a small footprint. The only reason I left this in was to keep it close to the original implementation [here](https://github.com/Azure/autorest-clientruntime-for-java/blob/v2/azure-client-authentication/src/main/java/com/microsoft/azure/v2/credentials/MSICredentials.java#L103)

Reply:
Completely agree that lib should have light weight footprint.  At the same time I'm not sure about exposing only sync variant to retrieve the token. We have had customers asking for full async support including the authentication step, this is why we introduced [AsyncServiceClientCredentials.](https://github.com/Azure/autorest-clientruntime-for-java/blob/v2/azure-client-runtime/src/main/java/com/microsoft/azure/v2/credentials/AsyncServiceClientCredentials.java) We might want to think how to handle it. 

We can always expose sync and let customer use rx/reactor construct to convert that to async. one problem I see is - customer may opt for io thread for this async conversion thinking token retrival is an nw io operation. But our retry strategy has back-off that can put this io thread in sleep mode, which is not recommend.